### PR TITLE
A Build & Test Workflow for GitHub Actions

### PR DESCRIPTION
**Pull request by simpsont-oci**

### CODE REVIEW

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -0,0 +1,899 @@
...
+        report_paths: OpenDDS/workspace/Tests_JUnit.xml
+
+  # Windows
+
+  windows-2019_defaults_artifact:
+
+    runs-on: windows-2019
+
+    steps:
+    - name: install openssl-windows & xerces-c
```
Comment (jwillemsen) [inline .github/workflows/build_and_test.yml:314]: Would recommend to use run-vcpkg, that uses caching, as far as I know the default vcpkg doesn't do that, see https://github.com/DOCGroup/ACE_TAO/blob/a72619e2ffcb37bc821c76b1e30b6886c2664003/.github/workflows/windows.yml#L99

File: .github/workflows/build_and_test.yml [hunk 2]
```diff
@@ -0,0 +1,899 @@
...
+    - name: configure OpenDDS
+      shell: cmd
+      run: |
+        call "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Auxiliary\Build\vcvars64.bat"
+        cd OpenDDS
+        configure --no-tests --security --ace=%GITHUB_WORKSPACE%/ACE_TAO/ACE --tao=%GITHUB_WORKSPACE%/ACE_TAO/TAO --mpc=%GITHUB_WORKSPACE%/MPC --xerces3=C:\vcpkg\installed\x64-windows --openssl=C:\vcpkg\installed\x64-windows
+    - name: build OpenDDS
+      shell: cmd
+      run: |
+        call "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Auxiliary\Build\vcvars64.bat"
```
Comment (jwillemsen) [inline .github/workflows/build_and_test.yml:359]: Easier is msvc-dev-cmd, see https://github.com/RemedyIT/taox11/blob/bbf9ec8d864aae68e77d9107a84cbd02f974a054/.github/workflows/windows.yml#L57

File: .github/workflows/build_and_test.yml [hunk 3]
```diff
@@ -0,0 +1,899 @@
...
+    - name: check build configuration
+      shell: cmd
+      run: |
+        call "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Auxiliary\Build\vcvars64.bat"
+        cd OpenDDS
+        call setenv.cmd
+        tools\scripts\show_build_config.pl
+    - name: build OpenDDS
+      shell: cmd
+      run: |
```
Comment (jwillemsen) [inline .github/workflows/build_and_test.yml:549]: Or use setup-msbuild, see https://github.com/DOCGroup/ACE_TAO/blob/a72619e2ffcb37bc821c76b1e30b6886c2664003/.github/workflows/windows.yml#L131

**Comment by jwillemsen** [comment]
Also MinGW64 can now be used on github actions, see https://github.com/DOCGroup/ACE_TAO/blob/c62ecd5955c1350fa25f95deb9bce2c81bc70f1c/.github/workflows/windows.yml#L137

**Comment by mitza-oci** [comment]
> 
> 
> Also MinGW64 can now be used on github actions, see https://github.com/DOCGroup/ACE_TAO/blob/c62ecd5955c1350fa25f95deb9bce2c81bc70f1c/.github/workflows/windows.yml#L137

MinGW (last time we checked) doesn't work.  So I suggest we not expand platform support as part of this effort.

File: .github/workflows/build_and_test.yml [hunk 4]
```diff
@@ -0,0 +1,908 @@
...
+      uses: mikepenz/action-junit-report@v1
+      if: always()
+      with:
+        github_token: ${{ secrets.GITHUB_TOKEN }}
+        check_name: "Test Results: macos-10_15_security"
+        report_paths: OpenDDS/workspace/Tests_JUnit.xml
+
+  macos-10_15_java_no-bit:
+
+    runs-on: macos-10.15
```
Comment (jwillemsen) [inline .github/workflows/build_and_test.yml:780]: There is also a macos-11.0 environment, see https://github.com/actions/virtual-environments#available-environments

File: .github/workflows/build_and_test.yml (hunk 1 above)
Comment (simpsont-oci) [reply .github/workflows/build_and_test.yml:314]: I've tried it, and it looks like it had a cache miss for the case when the cache key was identical, so it really didn't change the overall time by much. Perhaps I just need to run the workflow again with the same cache key, since it looks like updating the cache is sometimes post-build step?

I am, however, glad that I read the documentation for actions/cache, which is probably what we want to be using rather than artifacts to copy ACE/TAO files between jobs in the workflow. This is especially true if it will let you pull cached samples between workflow runs and lets you pull the cached values from master/main, which is how I understand the documentation. So I'll be looking into using that to avoid having to build the artifacts at all as part of normal pull request runs, which was part of the motivation for setting it up this way.

Comment (jwillemsen) [reply .github/workflows/build_and_test.yml:314]: First time run-vcpkg it makes the cache, in next runs it saves time, see the ACE-TAO actions

File: .github/workflows/build_and_test.yml [hunk 5]
```diff
@@ -0,0 +1,907 @@
+name: "Build & Test"
+
+on:
+  push:
+    branches:
+      - master
+      - branch-DDS-3.15
+      - branch-DDS-3.14
+      - branch-DDS-3.13
+      - gh_workflow
```
Comment (iguessthislldo) [inline .github/workflows/build_and_test.yml:10]:
```suggestion
      - gh_workflow
    paths:
      - '**'
      - '!docs/**'
      - '!README.md'
      - '!INSTALL.md'
      - '!NEWS.md'
```

This makes it not run the workflow if those are the only files that changed. We can add more if needed.

File: .github/workflows/build_and_test.yml [hunk 6]
```diff
@@ -0,0 +1,907 @@
...
+  pull_request:
+    branches:
+      - master
+      - branch-DDS-3.15
+      - branch-DDS-3.14
+      - branch-DDS-3.13
+      - gh_workflow
```
Comment (iguessthislldo) [inline .github/workflows/build_and_test.yml:17]:
```suggestion
      - gh_workflow
    paths:
      - '**'
      - '!docs/**'
      - '!README.md'
      - '!INSTALL.md'
      - '!NEWS.md'
```

File: .github/workflows/build_and_test.yml [hunk 7]
```diff
@@ -1538,7 +1544,7 @@ jobs:
       uses: actions/checkout@v2
       with:
         repository: DOCGroup/ACE_TAO
-        ref: 61338c7 # Update to 6_5_13 when available
+        ref: 9eae9c11f1bcd07b48c74820d8cb4c67aca9ef23 # Update to 6_5_13 when available
```
Comment (iguessthislldo) [inline .github/workflows/build_and_test.yml:1541]:
```suggestion
        ref: ace6tao2
```

Comment (jwillemsen) [reply .github/workflows/build_and_test.yml:1541]: Or the  ACE+TAO-6_5_12 tag?

**Comment by simpsont-oci** [comment]
OSX 11.0 didn't exist yet.

File: .github/workflows/build_and_test.yml [hunk 8]
```diff
@@ -0,0 +1,2093 @@
...
+      with:
+        github_token: ${{ secrets.GITHUB_TOKEN }}
+        check_name: "Test Results: ubuntu-16_04_java_no-bit"
+        report_paths: OpenDDS/workspace/Tests_JUnit.xml
+
+  # Windows 2019
+
+  windows-2019_defaults_artifact:
+
+    runs-on: windows-2019
```
Comment (jwillemsen) [inline .github/workflows/build_and_test.yml:895]: why duplicate everything for windows-2016 or windows-2019, using a matrix there is less duplication, see for example https://github.com/RemedyIT/taox11/blob/454f313787885b345a26a3960bcb49abb76ec432/.github/workflows/windows.yml#L27

File: .github/workflows/build_and_test.yml (hunk 7 above)
Comment (simpsont-oci) [reply .github/workflows/build_and_test.yml:1541]: OSX 11.0 support didn't exist for these.

File: .github/workflows/build_and_test.yml [hunk 9]
```diff
@@ -78,7 +78,7 @@ jobs:
     - name: upload ACE_TAO artifact
       uses: actions/upload-artifact@v2
       with:
-        name: ubuntu_ace_artifact
+        name: ubuntu-20_04_defaults_ace_artifact
```
Comment (jwillemsen) [inline .github/workflows/build_and_test.yml:81]: Will it not work when you use the job information, something like {{ job.name }}, would make it much easier for maintenance. You can also refer to the properties another job, see https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_idname.

File: dds/InfoRepo/DCPSInfoRepo.mpc [hunk 10]
```diff
@@ -1,4 +1,4 @@
-project(*Lib): dcpslib, iortable, dcps_inforepodiscovery, DCPSInfoRepo_bits, imr_client, svc_utils, install_lib {
+project(*Lib): dcpslib, dcps_tcp, iortable, dcps_inforepodiscovery, DCPSInfoRepo_bits, imr_client, svc_utils, install_lib {
```
Comment (mitza-oci) [inline dds/InfoRepo/DCPSInfoRepo.mpc:1]: Is this change necessary?  I thought it was already covered by `DCPSInfoRepo_bits`.

Comment (simpsont-oci) [reply dds/InfoRepo/DCPSInfoRepo.mpc:1]: Hmm, I thought I reverted these...

File: .github/workflows/build_and_test.yml [hunk 11]
```diff
@@ -0,0 +1,2073 @@
+name: "Build & Test"
+
+on:
+  push:
+    branches:
+      - master
+      - branch-DDS-3.15
+      - branch-DDS-3.14
+      - branch-DDS-3.13
```
Comment (mitza-oci) [inline .github/workflows/build_and_test.yml:9]: Does it allow wildcard expressions?  We don't necessarily need to do it now but maybe next iteration of this script could use them.

File: .github/workflows/build_and_test.yml [hunk 12]
```diff
@@ -0,0 +1,2073 @@
...
+    - name: checkout OpenDDS
+      uses: actions/checkout@v2
+      with:
+        path: OpenDDS
+        submodules: true
+    - name: checkout ACE_TAO
+      uses: actions/checkout@v2
+      with:
+        repository: DOCGroup/ACE_TAO
+        ref: ACE+TAO-6_5_12
```
Comment (mitza-oci) [inline .github/workflows/build_and_test.yml:68]:
```suggestion
        ref: ace6tao2
```

**Review by mitza-oci — CHANGES_REQUESTED** [review]

**Review by mitza-oci — APPROVED** [review]
